### DISCUSSION TITLE
[enhancement] Feature request: Show more audio format infos

### DISCUSSION THREAD

**Discussion by nousernouser** [jcorporation/myMPD#1184] [Ideas]
myMPD version: [13.0.5]

[enhancement] Feature request: Show more audio format infos

The _Audio format_ area in the _Playback view_ show bits, sample rate and channels for the current track, not the track's bitrate. None of those infos are shown in the _playback bar_ at bottom.

I think it would be useful (at least for the many audiophiles who are interested in this information) if:

- also the track bitrate was included in the _Audio format_ info area in the _Playback view_;
- the audio format info (bits, sample rate and channels) together with the instant bitrate were shown also in the _playback bar_ at bottom.

Thank you
Regards

 - MPD version: [0.23.12]
 - Browser: Firefox 120.0

**Comment by jcorporation** [comment]
Does mpd provide this information?

**Reply by nousernouser** [reply]
Hi  Jürgen

Yes:
- bits, sample rate and channels are already shown in the _Audio format_ info area in the myMPD _Playback view_
- instantaneous bitrate in kbps for the current track plyaback is provided by MPD:    
see `bitrate`: instantaneous bitrate in kbps` in the **status** section at  https://mpd.readthedocs.io/en/latest/protocol.html#querying-mpd-s-status

My request is to show all them (bits, sample rate, channels and instantaneous bitrate) in the _playback bar_ at bottom (footer).
Regards

**Reply by jcorporation** [reply]
Have you an idea how this additional information should be displayed?

How should we deal with the current bitrate? It can change very frequently if mpd plays music files with vbr.

**Reply by nousernouser** [reply]
> Have you an idea how this additional information should be displayed?

In the playback bar at the footer... something as in the following examples:

Now it is:
![mympd-foot-0](https://github.com/jcorporation/myMPD/assets/35839588/7d1cda93-3c4e-4ef6-b764-0b70cfe791b4)

Example1:
![mympd-foot-1](https://github.com/jcorporation/myMPD/assets/35839588/6f8ecb61-fd8b-4bdb-b131-7c0810c27cf2)

Example 2:
![mympd-foot-2](https://github.com/jcorporation/myMPD/assets/35839588/a37d072c-59de-4f03-8c53-9569accc0652)

> How should we deal with the current bitrate? It can change very frequently if mpd plays music files with vbr.

Yes, it's true.

I had in mind graphical clients (like Cantata) that implement this useful functionality... This is why I hadn't thought that with a web client to show the instant bitrate there is the problem of avoiding reloading the page... 

A possible solution could be using a js like:

```
<script>
setInterval(displayBitrate, 1000);

function mpdBitrate() {
    // get instantaneous bitrate from mpd
    // I don't know how reading the bitrate from mpd happens...
    ...
    ...
    return bitrate
}

function displayBitrate() {
    // 'bitrate' here is the id of the html element showing it
    document.getElementById("bitrate").innerHTML = mpdBitrate();
}
</script>
```

(The interval above is 1000 milliseconds... May be different: it should be the same as the one with which mpd updates the data, or a multiple... I don't know how to estimate...)

I hope it helps... What do you think?

**Comment by jcorporation** [comment]
I like your mockup, especially example 1. The bigger counter element looks good.

Yes, we can trigger the update of the bitrate with a JavaScript timer, but this wakes up 2 threads from myMPD and MPD itself.

The better alternative is that myMPD itself polls the bitrate status and sends it via the websocket to the browser instance.

But I do not like both ideas. Waking up the event based idle loop each second is a waste of resources.

**Comment by nousernouser** [comment]
> I like your mockup, especially example 1. The bigger counter element looks good.

Nice. I'm glad to hear it.

> Yes, we can trigger the update of the bitrate with a JavaScript timer, but this wakes up 2 threads from myMPD and MPD itself.
> 
> The better alternative is that myMPD itself polls the bitrate status and sends it via the websocket to the browser instance.
> 
> But I do not like both ideas. Waking up the event based idle loop each second is a waste of resources.

I understand that adding threads is not a great solution.

However, consider:

- it would be enough that instantaneous bitrate could be update every 3 or 5 seconds

- if I understand correctly and I'm not mistaken, the myMPD backend continuously gets the `elapsed` value from the MPD status  (also continuously updating the progress bar): wouldn't it be possible to get `bitrate` together with `elapsed`, using the same thread?

Regards

**Comment by jcorporation** [comment]
> if I understand correctly and I'm not mistaken, the myMPD backend continuously gets the elapsed value from the MPD status (also continuously updating the progress bar): wouldn't it be possible to get bitrate together with elapsed, using the same thread?

That is not correct, the elapsed time in the frontend is updated by a javascript timer.

**Comment by jcorporation** [comment]
There is a interesting pull request in the MPD repository that could help implementing the current bitrate future (and more): https://github.com/MusicPlayerDaemon/MPD/pull/1910

**Comment by jcorporation** [comment]
![image](https://github.com/jcorporation/myMPD/assets/765844/4506e883-3ae0-4bbd-91ba-293189710ac1)

In myMPD v14.0.0 is now the option to display the current audioformat in the footer. I will wait for the above MPD feature for the instant bitrate feature.

**Comment by nousernouser** [comment]
> In myMPD v14.0.0 is now the option to display the current audioformat in the footer. 

Very cool!

> I will wait for the above MPD feature for the instant bitrate feature.

I agree that it is the right way.

Thank you.
Regards

**Reply by GK-774** [reply]
I hope this is an option that can be switched off as well.

**Reply by jcorporation** [reply]
It is user configurable.
